subsys/debug/ppi_trace/ppi_trace.c

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <nrf_gpiote.h>
+#ifdef DPPI_PRESENT
+#include <nrfx_dppi.h>
+#else
+#include <nrfx_ppi.h>
+#endif
+#include <debug/ppi_trace.h>
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(ppi_trace, CONFIG_PPI_TRACE_LOG_LEVEL);
+
+/* Convert task address to associated subscribe register */
+#define SUBSCRIBE_ADDR(task) (volatile u32_t *)(task + 0x80)
+
+/* Convert event address to associated publish register */
+#define PUBLISH_ADDR(evt) (volatile u32_t *)(evt + 0x80)
+
+/* Handle which is used by the user to enable and disable the trace pin is
+ * encapsulating ppi channel(s). Bit 31 is set to indicate that handle is
+ * valid. Bit 30 is set to indicate that handle contains pair of PPI channels.
+ *
+ * Handle structure:
+ * ----------------------------------------------------------------
+ * |handle_flag|pair_flag|  not used | stop_ch | start_ch/main_ch |
+ * |-----------|---------|-----------|---------|------------------|
+ * |    31     |   30    |   29-16   | 15 - 8  |       7-0        |
+ * ----------------------------------------------------------------
+ */
+#define HANDLE_FLAG (1 << 31)
+#define PAIR_FLAG (1 << 30)
+
+#define HANDLE_ENCODE(value) (void *)((u32_t)value | HANDLE_FLAG)
+#define PACK_CHANNELS(start_ch, stop_ch) (PAIR_FLAG | start_ch | (stop_ch << 8))
+
+#define IS_VALID_HANDLE(handle) ((u32_t)handle & HANDLE_FLAG)
+#define IS_PAIR(handle) ((u32_t)handle & PAIR_FLAG)
+
+#define GET_CH(handle) ((u32_t)handle & 0xFF)
+#define GET_START_CH(handle) GET_CH(handle)
+#define GET_STOP_CH(handle) (((u32_t)handle >> 8) & 0xFF)
+
+/** @brief Allocate (D)PPI channel. */
+static nrfx_err_t ppi_alloc(u8_t *ch, u32_t evt)
+{
+	nrfx_err_t err;
+#ifdef DPPI_PRESENT
+	if (*PUBLISH_ADDR(evt) != 0) {
+		*ch = *PUBLISH_ADDR(evt) & 0x0F;
```

Comment:
This `& 0x0F` limits the number of supported channels to 16. What if we'll have more in some future chips?
I think it would be better to use masks from some arbitrary selected PUBLISH and SUBSCRIBE registers, and use them instead of these hard-coded magic bit masks (here and in `ppi_assign`).